discretize/tests.py

```diff
@@ -633,6 +637,121 @@ def Quadratic(x, return_g=True, return_H=True):
     return Quadratic
 
 
+def assert_isadjoint(
+    forward,
+    adjoint,
+    shape_u,
+    shape_v,
+    complex_u=False,
+    complex_v=False,
+    clinear=True,
+    rtol=1e-6,
+    atol=0.0,
+    assert_error=True,
+):
+    r"""Dot product test for the forward operator and its adjoint operator.
+
+    Dot product test to verify the correctness of the adjoint operator
+    :math:`F^H` of the forward operator :math:`F`.
+
+    .. math::
+
+        \mathbf{v}^H ( \mathbf{F} \mathbf{u} ) =
+        ( \mathbf{F}^H \mathbf{v} )^H \mathbf{u}
+
+
+    Parameters
+    ----------
+    forward : function
+        Forward operator.
+
+    adjoint : function
+        Adjoint operator.
+
+    shape_u : int, tuple of int
+        Shape of vector ``u`` passed in to ``forward``; it is accordingly the
+        expected shape of the vector returned from the ``adjoint``.
+
+    shape_v : int, tuple of int
+        Shape of vector ``v`` passed in to ``adjoint``; it is accordingly the
+        expected shape of the vector returned from the ``forward``.
+
+    complex_u : bool, default: False
+        If True, vector ``u`` passed to ``forward`` is a complex vector;
+        accordingly the ``adjoint`` is expected to return a complex vector.
+
+    complex_v : bool, default: False
+        If True, vector ``v`` passed to ``adjoint`` is a complex vector;
+        accordingly the ``forward`` is expected to return a complex vector.
+
+    clinear : bool, default: True
+        If operator is complex-linear (True) or real-linear (False).
+
+    rtol : float, default: 1e-6
+        Relative tolerance.
+
+    atol : float, default: 0.0
+        Absolute tolerance.
+
+    assert_error : bool, default: True
+        By default this test is an assertion (silent if passed, raising an
+        assertion error if failed). If set to False, the result of the test is
+        returned as boolean and a message is printed.
+
+
+    Returns
+    -------
+    passed : bool, returned if assert_error=False
```

Comment:
this line should only have the dtype with it, put any descriptions on the next line

```
passed : bool, optional
    Result of the dot product test. returned if `assert_error` is False.
```

Reply:
OK